allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+    throw it away.
+    """
+    origin = getattr(annotation, '__origin__', None)
+    args = getattr(annotation, '__args__', ())
+    if origin == Union and len(args) == 2 and args[1] == type(None):
+        return args[0]
+    else:
+        return annotation
+
+def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
+    """
+    Given some class, a `Params` object, and potentially other keyword arguments,
+    create a dict of keyword args suitable for passing to the class's constructor.
+    """
+    # Get the signature of the constructor.
+    signature = inspect.signature(cls.__init__)
+    kwargs: Dict[str, Any] = {}
+
+    # Iterate over all the constructor parameters and their annotations.
+    for name, param in signature.parameters.items():
+        # Skip "self". You're not *required* to call the first parameter "self",
+        # so in theory this logic is fragile, but if you don't call the self parameter
+        # "self" you kind of deserve what happens.
+        if name == "self":
+            continue
+
+        # Get the annotation and break it into `origin` and `args`
+        annotation = remove_optional(param.annotation)
+        origin = getattr(annotation, '__origin__', None)
+        args = getattr(annotation, '__args__', [])
+        default = param.default
+        optional = default != _NO_DEFAULT
+
+        # Some constructors expect extra non-parameter items, e.g. vocab: Vocabulary.
+        # We check the provided `extras` for these and just use them if they exist.
+        if name in extras:
+            kwargs[name] = extras[name]
+
+        # The next case is when the parameter type is itself constructible from_params.
+        elif hasattr(annotation, 'from_params'):
+            if name in params:
+                # Our params have an entry for this, so we use that.
+                subparams = params.pop(name)
+
+                if takes_arg(annotation.from_params, 'extras'):
+                    # If annotation.params accepts **extras, we need to pass them all along.
+                    subextras = extras
+                else:
+                    # Otherwise, only supply the ones that are actual args; any additional ones
+                    # will cause a TypeError.
+                    subextras = {k: v for k, v in extras.items() if takes_arg(annotation.from_params, k)}
+
+                # In some cases we allow a string instead of a param dict, so
+                # we need to handle that case separately.
+                if isinstance(subparams, str):
+                    kwargs[name] = annotation.by_name(subparams)()
+                else:
+                    kwargs[name] = annotation.from_params(params=subparams, **subextras)
+            elif not optional:
+                # Not optional and not supplied, that's an error!
+                raise ConfigurationError(f"expected key {name} for {cls.__name__}")
```

Comment:
Don't you need an `else` in here to use the default value?

Reply:
yes